### PR TITLE
19324 dropdownlist selection for boolean attributes

### PR DESCRIPTION
**Pull request by adnandautovic**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | #19324

@WinterSilence Just one question about the order of `$strict`: Should the comparison with `==` really be the strict one? Comparing with `strcmp` appears more strict to me, but for now I implemented your original suggestion. What do you think?

### CODE REVIEW

**Comment by adnandautovic** [comment]
> Comparing with `strcmp` appears more strict to me, but for now I implemented your original suggestion. What do you think?

In accordance with how `yii\helpers\BaseArrayHelper::isIn` handles `$strict`, I have now inverted the original proposal. Now, for `$strict = false`, the comparison is done with `==`.

**Comment by WinterSilence** [comment]
@adnandautovic [my solution](https://github.com/yiisoft/yii2/issues/19324#issuecomment-1084539935) works differently i.e. I not co-author

**Comment by WinterSilence** [comment]
@adnandautovic please add more test cases, for example, case from your issue: `renderSelectOptions(false, ['No', 'Yes'], ['prompt' => 'Please select'])`

**Comment by WinterSilence** [comment]
`ArrayHelper::isIn((string)$key, $selection, $strict)` this typing is incorrect: same result in strict and non-strict mode.

**Comment by adnandautovic** [comment]
> `ArrayHelper::isIn((string)$key, $selection, $strict)` this typing is incorrect: same result in strict and non-strict mode.

That's true, thank you. I took a look at this and it was caused by previous fixes for wrong default selections in d39d3ed0a65d5e442908f38f105fc6d3e3e01c2f and in e49a52f3a18dd7ae20a710dc8208baaab8a93fa8.
In fact, it appears that the fix in e49a52f3a18dd7ae20a710dc8208baaab8a93fa8 was faulty and the casting should have been removed in that commit.

Unfortunately, in order to do this properly then, I would have to change all three functions, `checkboxList`, `radioList` and `renderSelectOptions`.

I will see how I get along with this, but that definitely expands the scope of this issue.

**Comment by WinterSilence** [comment]
@adnandautovic invalid order array: ['Please select', No, Yes]

**Comment by adnandautovic** [comment]
> @adnandautovic invalid order in array: ['Please select', No, Yes]

Sorry, I don't understand? This particular test is not failing. Old tests for dropDownList, radioList and checkBoxList are failing, leading me to believe that I should just leave everything other than `renderSelectOptions` as it is.

**Comment by WinterSilence** [comment]
@adnandautovic look to HTML returning from updated `BaseHtml::renderSelectOptions()`

**Comment by adnandautovic** [comment]
> @adnandautovic look to HTML returning from updated `BaseHtml::renderSelectOptions()`

Thank you, I think I will be able to take a look next week.

In the meantime, I think the conversion to `(string)` has something to do with https://github.com/yiisoft/yii2/blob/8f4a7b696aad1f59ae9378721924ea8a91d85382/framework/helpers/BaseHtml.php#L1874-L1876 but I will have to check in more detail next time.

**Comment by adnandautovic** [comment]
> @adnandautovic look to HTML returning from updated `BaseHtml::renderSelectOptions()`

I think the order is fine, actually.


> In the meantime, I think the conversion to `(string)` has something to do with
> 
> https://github.com/yiisoft/yii2/blob/8f4a7b696aad1f59ae9378721924ea8a91d85382/framework/helpers/BaseHtml.php#L1874-L1876

In hindsight, this is obvious from d39d3ed0a65d5e442908f38f105fc6d3e3e01c2f.

> `ArrayHelper::isIn((string)$key, $selection, $strict)` this typing is incorrect: same result in strict and non-strict mode.

This has turned out to not be the case (see the tests I added in 8f4a7b696aad1f59ae9378721924ea8a91d85382). The results in strict and non-strict mode are different with the type casting and in fact the type casting is required.


Anyway, I have now added a failing test to demonstrate the impossibility of performing a non-strict check for a traversable selection containing a boolean, i.e. `[false]`, without explicitly handling the case of a boolean selection in `renderSelectOptions`.

The way I see it, there are two options:
1. Ignore that, because apparently no one has had a need for that so far.
2. Add code to separately handle the case where `$selection` contains a boolean.

I prefer the first, but both would be fine by me.

**Comment by WinterSilence** [comment]
@adnandautovic nope, you still don't understand how work typing: your array `[true => 'Yes', false => 'No']` convert to `[1 => 'Yes', 0 => 'No']` i.e.  `['No', 'Yes']`. Your expected HTML is invalid:
~~~php
$expected = <<<'EOD'
<option value="">Please select</option>
<option value="1">Yes</option>
<option value="0" selected>No</option>
EOD;
~~~

**Comment by adnandautovic** [comment]
> @adnandautovic nope, you still don't understand how work typing: your array `[true => 'Yes', false => 'No']` convert to `[1 => 'Yes', 0 => 'No']` i.e. `['No', 'Yes']`. Your expected HTML is invalid:
> 
> ```
> $expected = <<<'EOD'
> <option value="">Please select</option>
> <option value="1">Yes</option>
> <option value="0" selected>No</option>
> EOD;
> ```

As you can see from the tests that ran in the CI runner, the order is fine as far as php is concerned.
However, if you have some personal preference as to the order you want to see here, I can change that.

**Comment by adnandautovic** [comment]
> As you can see from the tests that ran in the CI runner, the order is fine as far as php is concerned. However, if you have some personal preference as to the order you want to see here, I can change that.

@WinterSilence for your convenience, I have commented out the one failing test that I added on purpose. As you can see, all tests run successfully now, even with the old html. But as I said, I can change it if you like.

**Comment by WinterSilence** [comment]
@adnandautovic Your current solution is valid, but work slow, better solution is fix expected HTML.

**Review by samdark — APPROVED** [review]

File: framework/CHANGELOG.md [hunk 1]
```diff
@@ -16,6 +16,7 @@ Yii Framework 2 Change Log
 - Enh #19304: Add filtering validator `yii\validators\TrimValidator` (WinterSilence)
 - Enh #19309: Optimize `yii\base\Model::attributes()` (WinterSilence)
 - Bug #19322: Revert force setting value to empty string in case it's `null` in `yii\validators\FilterValidator::validateAttribute()` (bizley)
+- Bug #19324: Fix yii\helpers\BaseHtml::renderSelectOptions() giving wrong selection for boolean attributes (adnandautovic)
```
Comment (bizley) [inline framework/CHANGELOG.md:19]:
```suggestion
- Bug #19324: Fix `yii\helpers\BaseHtml::renderSelectOptions()` giving wrong selection for boolean attributes (adnandautovic)
```

**Review by bizley — APPROVED** [review]

**Review by Arhell — APPROVED** [review]

**Comment by bizley** [comment]
Thank you.

**Pull request merged**